File: src/teleglance/parsing/message.py
"""Parsers for message nodes on t.me preview pages.

No class names live here — all structure comes from :mod:`.selectors`, so
markup drift is fixed by adjusting a :class:`Selectors` instance, not by
editing parser logic.
"""

from __future__ import annotations

from datetime import datetime
from functools import partial
from urllib.parse import parse_qs, urlparse

from selectolax.lexbor import LexborHTMLParser, LexborNode

from .._utils import bg_image_url, clean_text, parse_count
from ..errors import ParseError
from ..models import (
    DocumentRef,
    ForwardHeader,
    LinkPreview,
    Location,
    Media,
    Message,
    Photo,
    Poll,
    PollOption,
    Reaction,
    ReplyHeader,
    RoundVideo,
    Sticker,
    Video,
    Voice,
)
from .entities import extract_rich_text
from .registry import ParserRegistry
from .selectors import DEFAULT_SELECTORS, Selectors

# ---------------------------------------------------------------------------
# helpers


def _style_url(node: LexborNode | None) -> str | None:
    return bg_image_url(node.attributes.get("style")) if node is not None else None


def _classes(node: LexborNode) -> list[str]:
    return (node.attributes.get("class") or "").split()


def _has_ancestor_class(
    node: LexborNode, stop: LexborNode, classes: frozenset[str] | set[str]
) -> bool:
    parent = node.parent
    while parent is not None and parent is not stop:
        if any(c in classes for c in _classes(parent)):
            return True
        parent = parent.parent
    return False


def _msg_id_from_url(url: str | None) -> int | None:
    if not url:
        return None
    tail = url.rstrip("/").rsplit("/", 1)[-1].split("?", 1)[0]
    return int(tail) if tail.isdigit() else None


def _text_of(container: LexborNode, selector: str) -> str | None:
    node = container.css_first(selector)
    return clean_text(node.text(deep=True)) if node is not None else None


def _coordinates(url: str | None) -> tuple[float | None, float | None]:
    if not url:
        return None, None
    query = parse_qs(urlparse(url).query)
    value = next((query[key][0] for key in ("ll", "q") if query.get(key)), None)
    if not value or "," not in value:
        return None, None
    lat, lon = value.split(",", 1)
    try:
        return float(lat), float(lon)
    except ValueError:
        return None, None


# ---------------------------------------------------------------------------
# built-in block parsers — each takes the whole message node plus the active
# Selectors, and returns whatever media it recognizes


def parse_photos(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for wrap in node.css(sel.photo):
        url = _style_url(wrap)
        if url:
            out.append(Photo(url=url))
    return out


def parse_videos(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for player in node.css(sel.video_player):
        video = player.css_first(sel.video)
        video_classes = (video.attributes.get("class") or "") if video is not None else ""
        duration_node = player.css_first(sel.video_duration)
        duration = clean_text(duration_node.text()) if duration_node is not None else None
        url = video.attributes.get("src") if video is not None else None
        item: Media
        if sel.roundvideo_class in video_classes or player.css_first(sel.roundvideo) is not None:
            item = RoundVideo(
                url=url,
                thumb_url=_style_url(player.css_first(sel.roundvideo_thumb)),
                duration=duration,
            )
        else:
            item = Video(
                url=url,
                thumb_url=_style_url(player.css_first(sel.video_thumb)),
                duration=duration,
            )
        out.append(item)
    return out


def parse_voices(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for audio in node.css(sel.voice):
        src = audio.attributes.get("src")
        if not src:
            continue
        duration_node = node.css_first(sel.voice_duration)
        out.append(
            Voice(
                url=src,
                duration=clean_text(duration_node.text()) if duration_node is not None else None,
            )
        )
    return out


def parse_documents(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for doc in node.css(sel.document):
        title = _text_of(doc, sel.document_title)
        if title is None:
            continue
        out.append(DocumentRef(title=title, extra=_text_of(doc, sel.document_extra)))
    return out


def parse_stickers(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for sticker in node.css(sel.sticker):
        if sel.sticker_wrap_class in _classes(sticker):
            inner = sticker.css_first(sel.sticker_image)
            target = inner if inner is not None else sticker
        elif _has_ancestor_class(sticker, node, {sel.sticker_wrap_class}):
            continue  # already handled through its wrap
        else:
            target = sticker
        url = (
            target.attributes.get(sel.sticker_webp_attr)
            or _style_url(target)
            or (target.attributes.get("src") if target.tag == "img" else None)
        )
        out.append(Sticker(url=url, alt=target.attributes.get(sel.sticker_emoji_attr)))
    return out


def parse_polls(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for poll in node.css(sel.poll):
        options = []
        for option in poll.css(sel.poll_option):
            percent = option.css_first(sel.poll_option_percent)
            options.append(
                PollOption(
                    text=_text_of(option, sel.poll_option_text) or "",
                    percent=parse_count(percent.text()) if percent is not None else None,
                )
            )
        out.append(
            Poll(
                question=_text_of(poll, sel.poll_question) or "",
                kind=_text_of(poll, sel.poll_kind),
                options=options,
                voters=_text_of(node, sel.poll_voters),
            )
        )
    return out


def parse_link_previews(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for preview in node.css(sel.link_preview):
        out.append(
            LinkPreview(
                url=preview.attributes.get("href") or "",
                site_name=_text_of(preview, sel.link_preview_site_name),
                title=_text_of(preview, sel.link_preview_title),
                description=_text_of(preview, sel.link_preview_description),
                image_url=_style_url(preview.css_first(sel.link_preview_image)),
            )
        )
    return out


def parse_locations(node: LexborNode, sel: Selectors) -> list[Media]:
    out: list[Media] = []
    for wrap in node.css(sel.location):
        url = wrap.attributes.get("href")
        latitude, longitude = _coordinates(url)
        out.append(
            Location(
                url=url,
                image_url=_style_url(wrap.css_first(sel.location_image)),
                latitude=latitude,
                longitude=longitude,
            )
        )
    return out


def default_registry(selectors: Selectors | None = None) -> ParserRegistry:
    """A fresh registry with all built-in block parsers bound to the given
    selectors (defaults to :data:`DEFAULT_SELECTORS`)."""
    sel = selectors or DEFAULT_SELECTORS
    registry = ParserRegistry()
    registry.register("photo", partial(parse_photos, sel=sel))
    registry.register("video", partial(parse_videos, sel=sel))
    registry.register("voice", partial(parse_voices, sel=sel))
    registry.register("document", partial(parse_documents, sel=sel))
    registry.register("sticker", partial(parse_stickers, sel=sel))
    registry.register("poll", partial(parse_polls, sel=sel))
    registry.register("link_preview", partial(parse_link_previews, sel=sel))
    registry.register("location", partial(parse_locations, sel=sel))
    return registry


# ---------------------------------------------------------------------------
# message + feed parsing


def parse_message(
    node: LexborNode,
    registry: ParserRegistry,
    selectors: Selectors | None = None,
) -> Message | None:
    """Parse one message node. Returns None for nodes that carry no
    addressable post (service messages without a post id)."""
    sel = selectors or DEFAULT_SELECTORS

    data_post = node.attributes.get(sel.post_attr)
    if not data_post or "/" not in data_post:
        return None
    channel, _, msg_id_str = data_post.rpartition("/")
    if not msg_id_str.isdigit():
        return None
    msg_id = int(msg_id_str)

    date = None
    time_node = node.css_first(sel.date_time)
    if time_node is not None:
        raw = time_node.attributes.get(sel.datetime_attr)
        if raw:
            try:
                date = datetime.fromisoformat(raw)
            except ValueError:
                date = None

    views_node = node.css_first(sel.views)
    views_str = clean_text(views_node.text()) if views_node is not None else None

    text = html = markdown = None
    entities = []
    for candidate in node.css(sel.text):
        if _has_ancestor_class(candidate, node, sel.text_excluded_ancestor_classes):
            continue
        rich = extract_rich_text(candidate)
        text, html, markdown, entities = rich.text, rich.html, rich.markdown, rich.entities
        break

    forwarded = None
    fwd_node = node.css_first(sel.forward_name)
    if fwd_node is not None:
        forwarded = ForwardHeader(
            name=clean_text(fwd_node.text(deep=True)) or "",
            url=fwd_node.attributes.get("href"),
        )

    reply = None
    reply_node = node.css_first(sel.reply)
    if reply_node is not None:
        reply_url = reply_node.attributes.get("href")
        reply = ReplyHeader(
            author=_text_of(reply_node, sel.reply_author),
            text=_text_of(reply_node, sel.reply_text),
            url=reply_url,
            msg_id=_msg_id_from_url(reply_url),
        )

    reactions = []
    for reaction_node in node.css(sel.reaction):
        emoji_node = reaction_node.css_first(sel.reaction_emoji)
        emoji = clean_text(emoji_node.text(deep=True)) if emoji_node is not None else None
        count_str = _text_of(reaction_node, sel.reaction_count)
        if emoji:
            reactions.append(
                Reaction(
                    emoji=emoji,
                    count=parse_count(count_str),
                    count_str=count_str,
                    custom_emoji_id=(
                        emoji_node.attributes.get("emoji-id") if emoji_node is not None else None
                    ),
                )
            )

    comments_str = _text_of(node, sel.comments)

    return Message(
        id=msg_id,
        channel=channel,
        url=f"https://t.me/{channel}/{msg_id}",
        date=date,
        views=parse_count(views_str),
        views_str=views_str,
        author=_text_of(node, sel.author),
        text=text or "",
        html=html,
        markdown=markdown,
        entities=entities,
        media=registry.extract(node),
        forwarded_from=forwarded,
        reply_to=reply,
        edited=node.css_first(sel.edited) is not None,
        reactions=reactions,
        comments=parse_count(comments_str),
        comments_str=comments_str,
        raw_html=node.html or "",
    )


def parse_feed(
    html: str,
    registry: ParserRegistry,
    selectors: Selectors | None = None,
    *,
    strict: bool = False,
) -> list[Message]:
    """Parse a ``t.me/s/<channel>`` page (or fragment) into messages,
    in page order (oldest first)."""
    sel = selectors or DEFAULT_SELECTORS
    tree = LexborHTMLParser(html)
    messages = []
    candidates = 0
    for node in tree.css(sel.message):
        if sel.service_message_class in _classes(node):
            continue
        candidates += 1
        message = parse_message(node, registry, sel)
        if message is not None:
            messages.append(message)
    if strict and candidates and not messages:
        raise ParseError(f"found {candidates} message container(s), but none had a valid post id")
    return messages

File: packages/ts/src/parsing/message.ts
/**
 * Parsers for message nodes on t.me preview pages.
 *
 * No class names live here — all structure comes from Selectors, so
 * markup drift is fixed by adjusting a Selectors instance, not by
 * editing parser logic.
 */

import { parse, HTMLElement } from 'node-html-parser';
import { bgImageUrl, cleanText, parseCount } from '../utils.js';
import { ParseError } from '../errors.js';
import type {
  Message,
  ForwardHeader,
  ReplyHeader,
  Reaction,
} from '../models/message.js';
import type {
  Media,
  PollOption,
} from '../models/media.js';
import { extractRichText } from './entities.js';
import { ParserRegistry } from './registry.js';
import { DEFAULT_SELECTORS, type Selectors } from './selectors.js';

// ---------------------------------------------------------------------------
// helpers

function styleUrl(node: HTMLElement | null): string | null {
  return node ? bgImageUrl(node.getAttribute('style')) : null;
}

function classes(node: HTMLElement): string[] {
  const classValue = node.getAttribute('class') || '';
  return classValue.split(/\s+/).filter(Boolean);
}

function hasAncestorClass(
  node: HTMLElement,
  stop: HTMLElement,
  classSet: Set<string>,
): boolean {
  let parent = node.parentNode;
  while (parent && parent !== stop && parent instanceof HTMLElement) {
    const classValue = parent.getAttribute('class') || '';
    const parentClasses = classValue.split(/\s+/).filter(Boolean);
    if (parentClasses.some((c: string) => classSet.has(c))) {
      return true;
    }
    parent = parent.parentNode;
  }
  return false;
}

function msgIdFromUrl(url: string | null): number | null {
  if (!url) return null;
  const tail = url.replace(/\/$/, '').split('/').pop()?.split('?')[0];
  return tail && /^\d+$/.test(tail) ? parseInt(tail, 10) : null;
}

function textOf(container: HTMLElement, selector: string): string | null {
  const node = container.querySelector(selector);
  return node ? cleanText(node.textContent) : null;
}

function coordinates(url: string | null): [number | null, number | null] {
  if (!url) return [null, null];
  const parsed = new URL(url, 'https://t.me');
  const value = parsed.searchParams.get('ll') || parsed.searchParams.get('q') || '';
  if (!value || !value.includes(',')) return [null, null];
  const [lat, lon] = value.split(',', 2);
  const latNum = parseFloat(lat || '');
  const lonNum = parseFloat(lon || '');
  return [isNaN(latNum) ? null : latNum, isNaN(lonNum) ? null : lonNum];
}

// ---------------------------------------------------------------------------
// built-in block parsers — each takes the whole message node plus the active
// Selectors, and returns whatever media it recognizes

export function parsePhotos(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const wrap of node.querySelectorAll(sel.photo)) {
    const url = styleUrl(wrap as HTMLElement);
    if (url) {
      out.push({ type: 'photo', url });
    }
  }
  return out;
}

export function parseVideos(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const player of node.querySelectorAll(sel.videoPlayer)) {
    const video = player.querySelector(sel.video) as HTMLElement | null;
    const videoClasses = video ? (video.getAttribute('class') || '') : '';
    const durationNode = player.querySelector(sel.videoDuration);
    const duration = durationNode ? cleanText(durationNode.textContent || '') : null;
    const url = video?.getAttribute('src') || null;

    const isRound =
      videoClasses.includes(sel.roundvideoClass) ||
      player.querySelector(sel.roundvideo) !== null;

    if (isRound) {
      out.push({
        type: 'roundVideo',
        url,
        thumbUrl: styleUrl(player.querySelector(sel.roundvideoThumb) as HTMLElement),
        duration,
      });
    } else {
      out.push({
        type: 'video',
        url,
        thumbUrl: styleUrl(player.querySelector(sel.videoThumb) as HTMLElement),
        duration,
      });
    }
  }
  return out;
}

export function parseVoices(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const audio of node.querySelectorAll(sel.voice)) {
    const src = audio.getAttribute('src');
    if (!src) continue;
    const durationNode = node.querySelector(sel.voiceDuration);
    out.push({
      type: 'voice',
      url: src,
      duration: durationNode ? cleanText(durationNode.textContent) : null,
    });
  }
  return out;
}

export function parseDocuments(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const doc of node.querySelectorAll(sel.document)) {
    const title = textOf(doc as HTMLElement, sel.documentTitle);
    if (!title) continue;
    out.push({
      type: 'document',
      title,
      extra: textOf(doc as HTMLElement, sel.documentExtra),
    });
  }
  return out;
}

export function parseStickers(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const sticker of node.querySelectorAll(sel.sticker)) {
    const elem = sticker as HTMLElement;
    let target: HTMLElement;

    if (classes(elem).includes(sel.stickerWrapClass)) {
      const inner = elem.querySelector(sel.stickerImage) as HTMLElement;
      target = inner || elem;
    } else if (hasAncestorClass(elem, node, new Set([sel.stickerWrapClass]))) {
      continue; // already handled through its wrap
    } else {
      target = elem;
    }

    const url =
      target.getAttribute(sel.stickerWebpAttr) ||
      styleUrl(target) ||
      (target.tagName.toLowerCase() === 'img' ? target.getAttribute('src') : null);

    out.push({
      type: 'sticker',
      url: url || null,
      alt: target.getAttribute(sel.stickerEmojiAttr) || null,
    });
  }
  return out;
}

export function parsePolls(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const poll of node.querySelectorAll(sel.poll)) {
    const pollElem = poll as HTMLElement;
    const options: PollOption[] = [];
    for (const option of pollElem.querySelectorAll(sel.pollOption)) {
      const optionElem = option as HTMLElement;
      const percent = optionElem.querySelector(sel.pollOptionPercent);
      options.push({
        text: textOf(optionElem, sel.pollOptionText) || '',
        percent: percent ? parseCount(percent.textContent) : null,
      });
    }
    out.push({
      type: 'poll',
      question: textOf(pollElem, sel.pollQuestion) || '',
      kind: textOf(pollElem, sel.pollKind) || null,
      options,
      voters: textOf(node, sel.pollVoters),
    });
  }
  return out;
}

export function parseLinkPreviews(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const preview of node.querySelectorAll(sel.linkPreview)) {
    const previewElem = preview as HTMLElement;
    out.push({
      type: 'linkPreview',
      url: previewElem.getAttribute('href') || '',
      siteName: textOf(previewElem, sel.linkPreviewSiteName),
      title: textOf(previewElem, sel.linkPreviewTitle),
      description: textOf(previewElem, sel.linkPreviewDescription),
      imageUrl: styleUrl(
        previewElem.querySelector(sel.linkPreviewImage) as HTMLElement,
      ),
    });
  }
  return out;
}

export function parseLocations(node: HTMLElement, sel: Selectors): Media[] {
  const out: Media[] = [];
  for (const wrap of node.querySelectorAll(sel.location)) {
    const wrapElem = wrap as HTMLElement;
    const url = wrapElem.getAttribute('href');
    const [latitude, longitude] = coordinates(url || null);
    out.push({
      type: 'location',
      url: url || '',
      imageUrl: styleUrl(wrapElem.querySelector(sel.locationImage) as HTMLElement),
      latitude,
      longitude,
    });
  }
  return out;
}

/**
 * A fresh registry with all built-in block parsers bound to the given
 * selectors (defaults to DEFAULT_SELECTORS).
 */
export function defaultRegistry(selectors: Selectors = DEFAULT_SELECTORS): ParserRegistry {
  const registry = new ParserRegistry();
  const sel = selectors;

  registry.register('photo', (node: HTMLElement) => parsePhotos(node, sel));
  registry.register('video', (node: HTMLElement) => parseVideos(node, sel));
  registry.register('voice', (node: HTMLElement) => parseVoices(node, sel));
  registry.register('document', (node: HTMLElement) => parseDocuments(node, sel));
  registry.register('sticker', (node: HTMLElement) => parseStickers(node, sel));
  registry.register('poll', (node: HTMLElement) => parsePolls(node, sel));
  registry.register('link_preview', (node: HTMLElement) =>
    parseLinkPreviews(node, sel),
  );
  registry.register('location', (node: HTMLElement) => parseLocations(node, sel));

  return registry;
}

// ---------------------------------------------------------------------------
// message + feed parsing

/**
 * Parse one message node. Returns null for nodes that carry no
 * addressable post (service messages without a post id).
 */
export function parseMessage(
  node: HTMLElement,
  registry: ParserRegistry,
  selectors: Selectors = DEFAULT_SELECTORS,
): Message | null {
  const sel = selectors;

  const dataPost = node.getAttribute(sel.postAttr);
  if (!dataPost || !dataPost.includes('/')) {
    return null;
  }
  const [channel, msgIdStr] = dataPost.split('/');
  if (!msgIdStr || !/^\d+$/.test(msgIdStr)) {
    return null;
  }
  const msgId = parseInt(msgIdStr, 10);

  let date: Date | null = null;
  const timeNode = node.querySelector(sel.dateTime);
  if (timeNode) {
    const raw = timeNode.getAttribute(sel.datetimeAttr);
    if (raw) {
      try {
        date = new Date(raw);
        if (isNaN(date.getTime())) date = null;
      } catch {
        date = null;
      }
    }
  }

  const viewsNode = node.querySelector(sel.views);
  const viewsStr = viewsNode ? cleanText(viewsNode.textContent) : null;

  let text = '';
  let html: string | null = null;
  let markdown: string | null = null;
  let entities: any[] = [];

  const textNodes = node.querySelectorAll(sel.text);
  for (const candidate of textNodes) {
    const candidateElem = candidate as HTMLElement;
    if (hasAncestorClass(candidateElem, node, sel.textExcludedAncestorClasses)) {
      continue;
    }
    const rich = extractRichText(candidateElem);
    text = rich.text;
    html = rich.html;
    markdown = rich.markdown;
    entities = rich.entities;
    break;
  }

  let forwardedFrom: ForwardHeader | null = null;
  const fwdNode = node.querySelector(sel.forwardName) as HTMLElement | null;
  if (fwdNode) {
    forwardedFrom = {
      name: cleanText(fwdNode.textContent) || '',
      url: fwdNode.getAttribute('href') || null,
    };
  }

  let replyTo: ReplyHeader | null = null;
  const replyNode = node.querySelector(sel.reply) as HTMLElement | null;
  if (replyNode) {
    const replyUrl = replyNode.getAttribute('href') || null;
    replyTo = {
      author: textOf(replyNode, sel.replyAuthor),
      text: textOf(replyNode, sel.replyText),
      url: replyUrl,
      msgId: msgIdFromUrl(replyUrl),
    };
  }

  const reactions: Reaction[] = [];
  for (const reactionNode of node.querySelectorAll(sel.reaction)) {
    const reactionElem = reactionNode as HTMLElement;
    const emojiNode = reactionElem.querySelector(sel.reactionEmoji) as HTMLElement | null;
    const emoji = emojiNode ? cleanText(emojiNode.textContent || '') : null;
    const countStr = textOf(reactionElem, sel.reactionCount);
    if (emoji) {
      reactions.push({
        emoji,
        count: parseCount(countStr),
        countStr,
        customEmojiId: emojiNode?.getAttribute('emoji-id') || null,
      });
    }
  }

  const commentsStr = textOf(node, sel.comments);

  return {
    id: msgId,
    channel: channel || '',
    url: `https://t.me/${channel}/${msgId}`,
    date,
    views: parseCount(viewsStr),
    viewsStr,
    author: textOf(node, sel.author),
    text,
    html,
    markdown,
    entities,
    media: registry.extract(node),
    forwardedFrom,
    replyTo,
    edited: node.querySelector(sel.edited) !== null,
    reactions,
    comments: parseCount(commentsStr),
    commentsStr,
    rawHtml: node.outerHTML || '',
  };
}

/**
 * Parse a t.me/s/<channel> page (or fragment) into messages,
 * in page order (oldest first).
 */
export function parseFeed(
  html: string,
  registry: ParserRegistry,
  selectors: Selectors = DEFAULT_SELECTORS,
  opts: { strict?: boolean } = {},
): Message[] {
  const sel = selectors;
  const tree = parse(html);
  const messages: Message[] = [];
  let candidates = 0;

  for (const node of tree.querySelectorAll(sel.message)) {
    const elem = node as HTMLElement;
    if (classes(elem).includes(sel.serviceMessageClass)) {
      continue;
    }
    candidates += 1;
    const message = parseMessage(elem, registry, sel);
    if (message) {
      messages.push(message);
    }
  }

  if (opts.strict && candidates > 0 && messages.length === 0) {
    throw new ParseError(
      `Found ${candidates} message container(s), but none had a valid post id`,
    );
  }

  return messages;
}
